Feature/mission attendance

Hyeon0706 opened the pull request:
# 💡 PR Summary - <!--{ 작업 내용 }-->
<!-- 어떤 작업에 대한 PR 인지 위 주석에 적어주세요 -->
사용자 출석 기능 개발

### 📝 Description

---
<!-- 어떤 작업을 했는지 간단하게 적어주세요 -->
작업 사항

1. 사용자 출석 반영
2. 출석 여부에 따른 로직 처리

### 🔧 Working Branch

---
<!-- 예시) feature/user -->
작업 브랜치

feature/mission-attendance

### 📖 Related Issues

---
<!-- 예시) #1 -->
이슈 번호
- close #36 


### 📚 Etc

---
<!-- 작업 중 특이사항이 생기면 적어주세요 -->
추후 진행 작업

- #73

Hyeon0706 committed:
feat: Economy 도메인 패키지 생성

Hyeon0706 committed:
refactor: Mission 엔티티 매핑 관계 변경

- Member와 Mission이 1:1 관계이기 때문에 ManyToOne에서 OneToOne으로 변경

Hyeon0706 committed:
feat: 최근 출석 날짜를 받는 DTO 작성

- DB에서 데이터를 받아오기 위한 객체
- 담고 있는 내용
	- 최근 접속 날짜

Hyeon0706 committed:
feat: Mission 엔티티 커스텀 쿼리 레포지토리 추가

Hyeon0706 committed:
feat: Mission 엔티티 레포지토리 추가

Hyeon0706 committed:
feat: Mission 엔티티 출석 일자 반영을 위한 Update 쿼리 작성

- setLastAccessDate
- MemberId에 해당하는 데이터의 LastAccessDate를 현재 날짜로 변경

Hyeon0706 committed:
feat: 출석 체크 비즈니스 로직 작성

- 현재 날짜와 비교
- 현재 날짜와 같지 않을 경우 현재 날짜 Update

Hyeon0706 committed:
style: Google Code Style 적용

IamJuho commented on src/main/java/org/gitvest/gitvestb/economy/repository/EconomyRepositoryImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+  @Override
+  public MissionAttendance getLastAccessDate(Long memberId) {
+    QMission mission = QMission.mission;
```
QMission과 같은 Q객체들은 import static으로 선언하는 것은 어떤가요?

Hyeon0706 replied:
좋은 의견 감사합니다! static으로 변경 해 놓겠습니다.

IamJuho commented on src/main/java/org/gitvest/gitvestb/economy/repository/EconomyRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+@Repository
+public interface EconomyRepository extends JpaRepository<Mission, Long>, EconomyRepositoryCustom {
+  MissionAttendance getLastAccessDate(Long memberId);
```
EconmonyRepositoryCustom에 동일한 메서드가 선언되어 있는데 여기에도 추가로 선언하신 이유가 있을까요?

Hyeon0706 replied:
처음에 작성했던 코드입니다. 수정하도록 하겠습니다!

IamJuho commented on src/main/java/org/gitvest/gitvestb/economy/repository/EconomyRepositoryImpl.java:
```diff
@@ -25,4 +27,15 @@ public MissionAttendance getLastAccessDate(Long memberId) {
   }
+
+  @Override
+  @Transactional
```
트랜잭션을 서비스 레이어에서 걸어주는 것은 어떤가요?
레포지토리 레이어에서 트랜잭션을 걸어주신 이유가 궁금합니다!

Hyeon0706 replied:
QueryDSL이 처음이라 참조했던 블로그를 따라 했습니다. 생각해보니까 말씀하신게 맞는 것 같네요. 감사합니다!

IamJuho commented on src/main/java/org/gitvest/gitvestb/economy/service/EconomyServiceImpl.java:
```diff
@@ -0,0 +1,26 @@
+    // TODO : 에러 처리
+
+    LocalDate cur = LocalDate.now();
+    MissionAttendance dto = economyRepository.getLastAccessDate(memberId);
```
dto가 아닌 다른 이름으로 변수명을 정하는 것은 어떨까요? 
dto만으로는 무슨 객체의 변수명인지 파악하지 못할 수도 있을 것 같아요.

Hyeon0706 replied:
DTO클래스명 그대로 사용할까요? 변수명 네이밍 컨벤션이 없어서 일단 이렇게 작성했습니다!!

IamJuho replied:
그것도 좋은 방법으로 보여요~!

IamJuho reviewed (commented):
수고하셨어요!
예외 처리와 같은 추가적인 작업 후 merge하는 것이 좋아 보여요!

Hyeon0706 commented:
예외 처리는 예외 코드를 정리한 후 적용해서 다시 PR 올리겠습니다! 리뷰 감사합니다 😄

Hyeon0706 committed:
refactor: 리뷰 내용 적용

- EconomyServiceImpl
	- Transactional 어노테이션 추가
EconomyRepositoryImpl
	- Transactional 어노테이션 삭제
	- static으로 Q클래스 import
- EconomyRepository
	- 중복 메서드 삭제

IamJuho commented on src/main/java/org/gitvest/gitvestb/economy/repository/EconomyRepositoryImpl.java:
```diff
@@ -0,0 +1,28 @@
+    QMission mission = QMission.mission;
+
+    return queryFactory
+        .select(Projections.constructor(
+            MissionAttendance.class,
+            mission.lastAccessDate
```
저는 DTO를 조회하기 위해서 Projections.constructor() 메서드를 사용하는 것이 가독성이 다소 떨어진다고 생각합니다.
그 대체 방법으로 [@QueryProjections](https://wbsf.tistory.com/entry/QueryDsl%EC%97%90%EC%84%9C-DTO%EB%A1%9C-%EC%A1%B0%ED%9A%8C%ED%95%A0-%EB%95%8C-QueryProjection-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0)를 사용할 수 있는데, 이 방법은 어떠신가요?

Hyeon0706 replied:
장단점 알아보고 적용 해보도록 하겠습니다!

sktkddn777 commented on src/main/java/org/gitvest/gitvestb/economy/entity/Mission.java:
```diff
@@ -34,7 +35,7 @@ public class Mission extends Base {
   private LocalDate lastCommitDate;
 
-  @ManyToOne(fetch = FetchType.LAZY)
+  @OneToOne(fetch = FetchType.LAZY)
```
member 랑 mission이랑 OneToOne관계인가요?
이유가 듣고싶어요~

Hyeon0706 replied:
member 한 명당 미션을 위한 테이블 한 개만 있으면 됩니다!
mission 테이블은 memberId를 FK로 하고 최근 접속 날짜, 최근 커밋 날짜를 저장합니다.
따라서 두 테이블은 OneToOne 관계입니다!
